crates/federate/src/worker.rs

```diff
@@ -137,148 +215,245 @@ impl InstanceWorker {
+      self.state.last_successful_published_time = next.published;
+    }
+
+    let save_state_every = chrono::Duration::from_std(SAVE_STATE_EVERY_TIME).expect("not negative");
+    if force_write || (Utc::now() - self.last_state_insert) > save_state_every {
+      self.save_and_send_state().await?;
+    }
+    Ok(())
+  }
+
+  async fn spawn_send_if_needed(
+    &mut self,
+    activity_id: ActivityId,
+    report: UnboundedSender<SendActivityResult>,
+  ) -> Result<()> {
+    let Some(ele) = get_activity_cached(&mut self.pool(), activity_id)
+      .await
+      .context("failed reading activity from db")?
+    else {
+      tracing::debug!("{}: {:?} does not exist", self.instance.domain, activity_id);
+      report.send(SendActivityResult::Success(SendSuccessInfo {
+        activity_id,
+        published: None,
+        was_skipped: true,
+      }))?;
+      return Ok(());
+    };
+    let activity = &ele.0;
+    let inbox_urls = self
+      .get_inbox_urls(activity)
+      .await
+      .context("failed figuring out inbox urls")?;
+    if inbox_urls.is_empty() {
+      tracing::debug!("{}: {:?} no inboxes", self.instance.domain, activity.id);
+      report.send(SendActivityResult::Success(SendSuccessInfo {
+        activity_id,
+        published: Some(activity.published),
+        was_skipped: true,
+      }))?;
+      return Ok(());
+    }
+    let inbox_urls = inbox_urls.into_iter().collect();
+    let initial_fail_count = self.state.fail_count;
+    let data = self.config.to_request_data();
+    let stop = self.stop.clone();
+    let domain = self.instance.domain.clone();
+    tokio::spawn(async move {
+      let mut report = report;
+      if let Err(e) = InstanceWorker::send_retry_loop(
+        &ele.0,
+        &ele.1,
+        inbox_urls,
+        &mut report,
+        initial_fail_count,
+        domain,
+        data,
+        stop,
+      )
+      .await
+      {
         tracing::warn!(
           "sending {} errored internally, skipping activity: {:?}",
           ele.0.ap_id,
           e
         );
+        report.send(SendActivityResult::Success(SendSuccessInfo {
+          activity_id,
+          published: None,
+          was_skipped: true,
+        })).ok();
       }
-      if self.stop.is_cancelled() {
-        return Ok(());
-      }
-      // send success!
-      self.state.last_successful_id = Some(id);
-      self.state.last_successful_published_time = Some(ele.0.published);
-      self.state.fail_count = 0;
-    }
+    });
     Ok(())
   }
 
   // this function will return successfully when (a) send succeeded or (b) worker cancelled
   // and will return an error if an internal error occurred (send errors cause an infinite loop)
   async fn send_retry_loop(
-    &mut self,
-    pool: &mut DbPool<'_>,
     activity: &SentActivity,
     object: &SharedInboxActivities,
+    inbox_urls: Vec<Url>,
+    report: &mut UnboundedSender<SendActivityResult>,
+    initial_fail_count: i32,
```

Comment:
it's not passed back and forth, it's only passed one way, from the worker to the send task. the issue with not passing it in explicitly would be that then the send_retry_loop function would access `self`, and thus the _whole_ InstanceWorker struct would have to be Send+Sync, which means wrapping everything in locks.

That's why i create those local variables before the lambda calling this

Reply:
Then you can share around an `Arc<AtomicI32>` instead of `self`.

Reply:
The point is we don't need any atomics, so adding atomics without any real reason is bad. this is a read-only immutable variable that is passed only one way, from the main thread to each send task so it knows if the request fails how much to sleep initially.

Reply:
just realizing you're referring to the channel send when you are talking about the "passing it back".

I see that might be an option. But I don't like it. The send task is cleanly separated and has no direct interaction with the main thread. It's nicely compartmentalized and you can (after the split) now purely look at the send.rs file to understand how the send-retry-loop of one activity works. It purely sends updates upwards, it never receives updates downwards.

The failure event needs to be transferred to the main task in any case so it can update the database. The main fail_count should only be updated under certain conditions and these conditions belong in the main task, since the logic that updates the database needs the same information.

In addition, if we changed this to an atomic for the fails then we should also change the "success" sends to direct writes into a RwLock<BinaryHeap> instead of writing to a channel. Otherwise there's a weird mix where some data is sent via channel and some via atomics. But changing that would mean you would again have to figure out how to signal the main task when to continue and again make all interactions more complex to understand since they can happen with arbitrary concurrency.

Reply:
Alright, it seems a bit weird at first but makes sense with your explanation. Would be good if you can copy that into a code comment on `SendActivityResult`.